articles/moricho.re

```diff
@@ -0,0 +1,523 @@
+また、@<code>{/sys/fs/cgroup/cpu/shoten}内でもう１つ重要なファイルが@<code>{tasks}です。このファイル内では、どのプロセスをこのcontrol groupに入れるかを管理しています。複数プロセス入れることもできますが、本項ではコンテナとなるプロセス自身のみを設定しましょう。ではcgroupのコードを見ていきます。
+
+//list[cgroup3][cgroupの実装][go]{
+  func cgroup() error {
+  	if err := os.MkdirAll("/sys/fs/cgroup/cpu/shoten", 0700); err != nil {
+  		return fmt.Errorf("failed to create directory: %w", err)
+  	}
+
+  	if err := ioutil.WriteFile(
+  		"/sys/fs/cgroup/cpu/shoten/tasks",
+  		[]byte(fmt.Sprintf("%d\n", os.Getpid())),
+  		0644,
+  	); err != nil {
+  		return fmt.Errorf("failed to register tasks: %w", err)
+  	}
+
+  	if err := ioutil.WriteFile(
+  		"/sys/fs/cgroup/cpu/shoten/cpu.cfs_quota_us",
+  		[]byte("5000\n"),
+  		0644,
+  	); err != nil {
+  		return fmt.Errorf("failed to limit cpu.cfs_quota_us: %w", err)
+  	}
+
+  	return nil
+  }
+//}
+
+とてもシンプルで、主に次の3つを行っています。
+
+ * @<code>{/sys/fs/cgroup/cpu/shoten}ディレクトリを作成
+ * @<code>{tasks}ファイルに自身のPIDを書き込み
+ * @<code>{cpu.cfs_quota_us}にCPU使用率が５％になるように5000という数字の書き込み
+
+ではまた@<code>{reexec}パッケージを使って、@<list>{cgroup3}を適用していきましょう。変更箇所は@<code>{InitContainer}関数だけです。
+
+//list[cgroup4][InitContainer()の更新][go]{
+func InitContainer() {
+  if err := cgroup(); err != nil {
+		fmt.Printf("Error running cgroup - %s\n", err)
+		os.Exit(1)
+	}
+
+	...
+
+  if err := pivotRoot(newrootPath); err != nil {
+		fmt.Printf("Error running pivot_root - %s\n", err)
+		os.Exit(1)
+	}
+
+	Run()
+}
+//}
+
+前節までの@<code>{pivotRoot}関数に加え、@<code>{cgroup}関数を追加しました。このとき@<code>{cgroup}関数と@<code>{pivotRoot}関数の位置関係に注意してください。@<code>{cgroup}関数ではホストの@<code>{/sys/fs/cgroup/cpu/shoten}配下にあるファイルへ書き込みを行いたいため、@<code>{pivotRoot}関数でルートファイルシステムを変更する前に実行しています。
+
+ではCPU使用率が５％に制限されているか確認してみましょう。まず@<code>{/bin/sh}が実行されたプロセス内で「@<code>{while :; do true ; done}」を実行し、CPUに負荷をかけます。@<code>{cgroups}の機能で制限していない場合、通常ならCPU使用率が100％近くになります。
+
+//list[cgroup5][負荷をかける][]{
+$ go build -o main
+$ ./main
+-[shoten]- # while :; do true ; done
+
+//}
+
+では別のセッションに入り@<code>{top}コマンドでCPU使用率を確認してみます。
+
+//list[cgroup6][別セッションでCPU使用率を確認][]{
+PID   USER     PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND
+20493 root     20   0    8100   2100   1988 R  5.0  0.2   0:02.31 sh
+20270 root     20   0  404360  18924  12840 S  0.3  1.9   0:02.06 ssm-session-wor
+1     root     20   0  225140   9212   7088 S  0.0  0.9   0:07.44 systemd
+2     root     20   0       0      0      0 S  0.0  0.0   0:00.00 kthreadd
+...
+//}
+
+一番上の@<code>{sh}を実行しているプロセスが該当するものです。見事に５％前後でキープされているのがわかります。
+
+== 自作コンテナに機能を追加する
+ここまでを通してコンテナの核となる機能を実装できました。お疲れ様です。本節では、ここからさらに機能拡張をする場合に何をすればいいか簡単に紹介します。
+
+==== vethとLinux Bridge
+「7.3 カーネルリソースの隔離」の@<list>{namespace1}では、@<code>{syscall.SysProcAttr}構造体の@<code>{Cloneflags}フィールドに@<tt>{CLONE_NEWNET}フラグをセットすることでネットワーク名前空間を分離しました。
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Line 501 sentence length(102) exceeds the maximum sentence length of 100.
Over 2 characters. (ja-technical-writing/sentence-length)